File: Othello-Src-Cpp/src/player.cpp
#include <game_util.h>
#include <player.h>

/*
    Default Constructor
*/
Player::Player() {
  this->is_cpu   = true;
  this->is_black = true;
}

/*
    Constructor with arguments
*/
Player::Player(bool _is_cpu, bool _is_black) {
  this->is_cpu   = _is_cpu;
  this->is_black = _is_black;
}

/*
    Set Color
*/
void Player::set_color(bool _is_black) { this->is_black = _is_black; }

/*
    Set Computer Played Unit
*/
void Player::set_cpu(bool _is_cpu) { this->is_cpu = _is_cpu; }

/*
    Get Step
*/
Coord Player::get_step(Board &_board) {
  Coord step = NO_ACTION;
  if (this->is_cpu) {
    // Do UCT Search
    // Color is the Last Piece's Color
    step = uct.uct_search(_board, !this->is_black);
  } else {
    // Ask User to Input
    while (true) {
      int r, c;
      printf("Please input valid step in format < r , c >: ");
      scanf("%d %d", &r, &c);
      step = Coord(r, c);
      if (GameUtil::is_valid_step(_board, this->is_black, step)) break;
    }
  }
  return step;
}

File: Othello-Src-Python/player.py
#!/usr/bin/env python3
import board as board
import coord as coord
import uct as uct
import game_util as game_util


class Player():
    def __init__(self, **kwargs):
        """ Constructor """
        for key in kwargs.keys():
            if key != 'is_cpu' and key != 'is_black':
                raise BaseException(
                    'Error Key Argument for Player Constructor')

        self.is_cpu = kwargs['is_cpu'] if 'is_cpu' in kwargs else True
        self.is_black = kwargs['is_black'] if 'is_black' in kwargs else True

        self.uct = uct.Tree()

    def __repr__(self):
        return self.__str__()

    def __str__(self):
        return "[CPU:"+self.is_cpu.__str__()+", IS_BLACK:"+self.is_black.__str__()+"]"

    def set_color(self, _is_black: bool) -> None:
        self.is_black = _is_black

    def set_cpu(self, _is_cpu: bool) -> None:
        self.is_cpu = _is_cpu

    def get_step(self, _board: board.Board) -> coord.Coord:
        step = coord.NO_ACTION
        if(self.is_cpu):
            # Do UCT Search
            # Color is the Last Pieces's Color
            step = self.uct.uct_search(_board, not (self.is_black))
        else:
            while(True):
                r, c = input(
                    'Please input valid step in format < r  c >: ').split()
                step = coord.Coord(r, c)
                if(game_util.is_valid_step(_board, self.is_black, step)):
                    break

        return step
